firedrake/mesh.py

```diff
@@ -1537,6 +1691,157 @@ def ExtrudedMesh(mesh, layers, layer_height=None, extrusion_type='uniform', kern
+
+    vmesh._callback = callback
+    return vmesh
+
+
+def _pic_swarm_in_plex(plex, coords, **kwargs):
+    """
+    Create a Particle In Cell (PIC) DMSwarm, immersed in a DMPlex
+    at given point coordinates.
+
+    This should only by used for dmplexes associated with meshes with
+    straight edges. If not, the particles may be placed in the wrong
+    cells.
+
+    :arg plex: the DMPlex within with the DMSwarm should be
+        immersed.
+    :arg coords: a list of point coordinate tuples at which to create
+        the particles.
+    :kwarg comm: Optional communicator to build the mesh on
+        (defaults to match `plex` communicator).
+    :kwarg fieldnames: An optional list of field names for fields to
+        register on the swarm.
+    :kwarg blocksizes: An optional list of blocksizes for fields to
+        register on the swarm. If specfied must bo of equal length to
+        `fieldnames`. If not specified, defaults to 1 for all specified
+        `fieldnames`.
+    :return: the immersed DMSwarm
+    """
+
+    comm = kwargs.get("comm", plex.comm)
+
+    coords = np.asarray(coords, dtype=np.double)
+
+    # Create a DMSWARM
+    swarm = PETSc.DMSwarm().create(comm=comm)
+
+    # Set swarm DM dimension to match DMPlex dimension
+    # NB: Unlike a DMPlex, this does not correspond to the topological
+    #     dimension of a mesh (which would be 0). In all PETSc examples
+    #     the dimension of the DMSwarm is set to match that of the
+    #     DMPlex used with swarm.setCellDM
+    swarm.setDimension(plex.getDimension())
+
+    # Set coordinates dimension
+    if len(np.shape(coords)) == 1:
+        coordsdim = 1
+    elif len(np.shape(coords)) == 2:
+        coordsdim = np.shape(coords)[1]
+    else:
+        raise ValueError("Point coordinates list shape unsupported")
+    swarm.setCoordinateDim(coordsdim)
+
+    if coordsdim == 1:
+        raise NotImplementedError("1D DMSwarm not yet supported")
+
+    # Link to DMPlex cells information for when swarm.migrate() is used
+    swarm.setCellDM(plex)
+
+    # Set to Particle In Cell (PIC) type
+    swarm.setType(PETSc.DMSwarm.Type.PIC)
+
+    # Register any fields
+    fieldnames = kwargs.get("fieldnames", [])
+    blocksizes = kwargs.get("blocksizes", np.ones(len(fieldnames)))
+    if len(fieldnames) > 0:
+        if len(fieldnames) != len(blocksizes):
+            raise ValueError("The blocksizes list must be the same length as the fieldnames list")
+        for i in range(len(fieldnames)):
+            swarm.registerField(fieldnames[i], blocksizes[i])
+    swarm.finalizeFieldRegister()
+
+    # Note that no new fields can now be associated with the DMSWARM.
+
+    # Add point coordinates - note we set redundant mode to False since
+    # all MPI ranks are given the same list of coordinates. This forces
+    # all ranks to search for the given coordinates within their cell.
+    swarm.setPointCoordinates(coords, redundant=False, mode=PETSc.InsertMode.INSERT_VALUES)
```

Comment:
I've clarified this in 1fa6c16d8c204f2168d17226738dbba6e516a30b. This does make me wonder if I need to somehow work out the "topological" coordinate of each vertex before using this function - I don't know how this will behave with immersed manifolds.

Reply:
Disabled support for immersed manifolds for now.